x/stake/keeper/query_utils.go

```diff
@@ -0,0 +1,173 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+// Return all validators that a delegator is bonded to. If maxRetrieve is supplied, the respective amount will be returned.
+func (k Keeper) GetDelegatorValidators(ctx sdk.Context, delegatorAddr sdk.AccAddress,
+	maxRetrieve ...int16) (validators []types.Validator) {
+
+	retrieve := len(maxRetrieve) > 0
+	if retrieve {
+		validators = make([]types.Validator, maxRetrieve[0])
+	}
+	store := ctx.KVStore(k.storeKey)
+	delegatorPrefixKey := GetDelegationsKey(delegatorAddr)
+	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) //smallest to largest
+	defer iterator.Close()
+
+	i := 0
+	for ; iterator.Valid() && (!retrieve || (retrieve && i < int(maxRetrieve[0]))); iterator.Next() {
+		addr := iterator.Key()
+		delegation := types.MustUnmarshalDelegation(k.cdc, addr, iterator.Value())
+		validator, found := k.GetValidator(ctx, delegation.ValidatorAddr)
+		if !found {
+			panic(types.ErrNoValidatorFound(types.DefaultCodespace))
+		}
+
+		if retrieve {
+			validators[i] = validator
+		} else {
+			validators = append(validators, validator)
+		}
+
+		i++
+	}
+	return validators[:i] // trim
+}
+
+// Return all validators that a delegator is bonded to. If maxRetrieve is supplied, the respective amount will be returned.
+func (k Keeper) GetDelegatorBechValidators(ctx sdk.Context, delegatorAddr sdk.AccAddress,
```

Comment:
I still don't see the need for these auxiliary methods? Can't I just call `keeper. GetDelegatorValidators` and then loop over each validator that result and call `Bech32Validator`. This will reduce code bloat.

Reply:
True -- but the second iteration is extremely trivial with very little to no overheads (talking ns). Overall it keeps the code very DRY.